src/examples/learn_quant/monotonicity.py

```diff
@@ -0,0 +1,215 @@
+from ultk.effcomm.informativity import informativity
+from ultk.language.grammar import GrammaticalExpression
+from ultk.language.language import Language, aggregate_expression_complexity
+from ultk.language.semantics import Meaning
+
+from learn_quant.meaning import create_universe
+from learn_quant.util import read_expressions
+from learn_quant.quantifier import QuantifierUniverse, QuantifierModel
+from concepts.contexts import Context
+
+from itertools import product
+import math
+from scipy.stats import entropy
+import pandas as pd
+import numpy as np
+from tqdm import tqdm
+from scipy import sparse
+
+from pathlib import Path
+
+import hydra
+from omegaconf import DictConfig, OmegaConf
+
+
+def switch_direction(bools: list[list[bool]], down = False) -> list[list[bool]]:
+        """Switches the value of True to False and vice versa.
+
+        Args:
+            bool (list[list[bool]]): Accepts a list of lists of bools 
+
+        Returns:
+            list[list[bool]]: A list of lists of bools, flipped unless self.down == True. 
+        """
+        if down:
+            return bools
+        else:
+            return np.invert(bools)
+        
+def get_truth_matrix(universe: QuantifierUniverse) -> pd.DataFrame:
+    """Get a pandas dataframe that stores the truth vectors of all models in the quantifier universe
+
+    Returns:
+        pd.DataFrame: A dataframe of booleans arrays that track the indices of set items of A intersected by B.
+    """
+    truth_array = []
+    for quantifier_model in universe.referents:
+        truth_vector = tuple(True if x == '2' else False for x in quantifier_model.name)
+        truth_array.append(truth_vector)
+    truth_df = pd.concat([pd.DataFrame(universe.get_names()).rename(columns={0: 'names'}), pd.DataFrame(truth_array)], axis=1).set_index("names")
+    return truth_df
+
+def calculate_lattice(universe: QuantifierUniverse, down=False) -> Context:
+        """ 
+
+        Returns:
+            Context: Created from ``objects``, ``properties``, and ``bools`` correspondence.
+
+            objects: Iterable of object label strings.
+            properties: Iterable of property label strings.
+            bools: Iterable of ``len(objects)`` tuples of ``len(properties)`` booleans.
+
+            See https://github.com/haberchr/concepts/blob/master/concepts/contexts.py
+
+        """
+        truth_df = get_truth_matrix(universe)
+        objects = truth_df.index.astype(str).tolist()
+        properties = [x for x in truth_df.columns]
+        bools = list(truth_df.fillna(False).astype(bool).itertuples(index=False, name=None))
```

Comment:
Let's talk about this in next meeting; am curious about whether this is better than NP version still